CMakeLists.txt

```diff
@@ -580,6 +580,21 @@ if(EXISTS ${APPLICATION_SOURCE_DIR}/VERSION)
   add_dependencies(zephyr_interface app_version_h)
 endif()
 
+# Create the kernel library now so that other parts of the system can use
+# it with target_link_libraries or other commands.
+#
+# If a pre-built static library containing kernel code exists in the kernel
+# directory, kernel/libkernel.a, link it with the application code instead
+# of building from source.
+
+set(libkernel ${CMAKE_CURRENT_SOURCE_DIR}/kernel/libkernel${CMAKE_STATIC_LIBRARY_SUFFIX})
+if(EXISTS ${libkernel})
+  add_library(kernel STATIC IMPORTED GLOBAL)
+  set_property(TARGET kernel PROPERTY IMPORTED_LOCATION ${libkernel})
+else()
+  add_library(kernel)
+endif()
```

Comment:
i'm currently cleaning up Zephyr link handling, so I would like to know more details on the background reasons for this move.

The commit message gives a small hint, but not enough for me to actually be able to judge if the upcomming build system improvements are actually solving the issues faced.

Can you provide some more details / examples on what exactly this move will solve ?

Reply:
This creates the kernel library early in the build process so that dependency information can be associated with it by early libraries in the system. In the follow-on patch, which actually makes the C library support bits into regular libraries, we need to construct a circular dependency graph including the kernel and C libraries. Doing that in the C library code instead of the kernel code seemed a bunch cleaner to me, but that required having the kernel library created before the C library support bits were processed by cmake.